README.md
# CPF

A library to work with CPFs.

[![Hex pm](https://img.shields.io/hexpm/v/cpf.svg?style=flat)](https://www.hex.pm/packages/cpf)
[![CircleCI](https://circleci.com/gh/ulissesalmeida/cpf/tree/master.svg?style=svg)](https://circleci.com/gh/ulissesalmeida/cpf/tree/master)
[![Coverage Status](https://coveralls.io/repos/github/ulissesalmeida/cpf/badge.svg?branch=master)](https://coveralls.io/github/ulissesalmeida/cpf?branch=master)

CPF is an acronym for "Cadastro de Pessoa Físicas," it's a unique number
associated to a person that the Brazilian government maintains. With this
number, it is possible to check or retrieve information about a person.

This library provides a validation that checks if the number is a valid CPF
number. The CPF has check digit algorithm is similar to
[`ISBN 10`](https://en.wikipedia.org/wiki/Check_digit#ISBN_10), you can check
the details in Portuguese [here](https://pt.wikipedia.org/wiki/Cadastro_de_pessoas_f%C3%ADsicas#C%C3%A1lculo_do_d%C3%ADgito_verificador).

## Installation

If [available in Hex](https://hex.pm/docs/publish), the package can be installed
by adding `cpf` to your list of dependencies in `mix.exs`:

```elixir
def deps do
  [
    {:cpf, "~> 1.2"}
  ]
end
```

## Quick Start

You can verify if a CPF is valid by calling the function `CPF.valid?/1`:

```elixir
iex> CPF.valid?(563_606_676_73)
true

iex> CPF.valid?(563_606_676_72)
false

iex> CPF.valid?("563.606.676-73")
true

iex> CPF.valid?("563.606.676-72")
false

iex> CPF.valid?("56360667673")
true

iex> CPF.valid?("56360667672")
false
```

## Parsing and Storing CPFs

If you want to store CPF as integer or as `String.t`, this library have
you covered. You can do:

```elixir
iex> "044.858.476-08" |> CPF.parse!() |> CPF.to_integer()
4485847608

iex> "044.858.476-08" |> CPF.parse!() |> to_string()
"04485847608"
```

Storing CPF as strings are easier for a human to read since the 0 padding digits
are there. Meanwhile, storing as integers will allow you have better performance
in CPF lookups.

The `CPF.parse/1` and `CPF.parse!/1` returns you the CPF value wrapped in a
custom type with explicit digits.

```elixir
iex> CPF.parse("044.858.476-08")
{:ok, #CPF<"044.858.476-08">}

iex> CPF.parse("044.858.476-07")
{:error, %CPF.ParsingError{reason: :invalid_verifier}}

iex> CPF.parse!("044.858.476-08")
#CPF<"044.858.476-08">

iex> CPF.parse!("044.858.476-07")
** (CPF.ParsingError) invalid verifier
```

With the casted CPF in hands, you can use `CPF.format/1`, `CPF.to_integer/1` and
`to_string/1`.

## CPF Formatting

If you have a valid CPF strings or integer in hands, you can use `CPF.new/1` and
in sequence call `CPF.format/1`:

```elixir
iex> 4485847608 |> CPF.new() |> CPF.format()
"044.858.476-08"

iex> "04485847608" |> CPF.new() |> CPF.format()
"044.858.476-08"
```

The `CPF.format/1` expects the input be wrapped in the CPF type. Remember, only
use `CPF.new` with valid CPFs, no other checks are done there. If you need some
validation, use `CPF.parse/1`.

## Generating random CPFs for testing

You can generate valid CPF numbers by using `CPF.generate/0`:

```elixir
iex> CPF.generate()
#CPF<671.835.731-68>

iex> CPF.generate() |> to_string()
"67183573168"

iex> CPF.generate() |> CPF.to_integer()
67183573168

iex> CPF.generate() |> CPF.format()
"671.835.731-68"
```

After you generate the CPF, you can turn the CPF into a formatted string, or
convert to a string digits, or convert to integer.

## Flexibilizing the CPF validation

You can use `CPF.flex/1` when you only care if the user has provided the
correct number before any validation or parsing. For example:

```elixir
iex> "04.4.8*58().476-08" |> CPF.flex() |> CPF.valid?()
true

iex> "04.4.8*58().476-08" |> CPF.flex() |> CPF.parse!() |> CPF.format()
"044.858.476-08"
```

It can be useful to take a user's dirty input and format it.

## Command line

You can generate random valid CPFs with:

```shell
$ mix cpf.gen
194.925.115-25

$ mix cpf.gen --format=digits --count=2
19492511525
65313188640
```

You can also check if CPF are valid with:

```shell
$ mix cpf.check 194.925.115-25
valid
```

Run `mix help cpf.gen` and `mix help cpf.check` to read a further explanation
about the commands available options.

## Ecto Integration

If you have `ecto` installed in your application, you can use the module
`CPF.Ecto.Type` or `CPF.Ecto.Changeset` to cast and validate CPF fields.


### Using `CPF.Ecto.Type`

If you like the strictness of types you can user `CPF.Ecto.Type` functions to
define in your schema the CPF field type. For example:

```elixir
defmodule MyApp.Profile do
  use Ecto.Schema

  import Ecto.Changeset
  # Import the parameterized type function
  import CPF.Ecto.Type

  schema "profiles" do
    # use `cpf_type/1` in the field type definition
    field :cpf, cpf_type(:string)
  end

  def new(enum \\ %{}), do: struct!(__MODULE__, enum)

  def changeset(profile, params), do: cast(profile, params, __schema__(:fields))
end
```

It will prevent any attempt to insert an invalid CPF when using your
schema module. For example:

```elixir
{:error, changeset} =
  Profile.new()
  |> Profile.changeset(%{cpf: "abilidebob"})
  |> MyApp.Repo.insert()

{"is invalid", [reason: :invalid_format]} = Keyword.get(changeset.errors, :cpf)
```

Note: it will also prevent queries with invalid CPFs by raising casting error.
You might need workaround it if you want to display an empty result with invalid
CPFs. For example:

```elixir
if CPF.valid?(cpf) do
  query = from p in Profile, where: p.cpf == ^cpf
  Repo.all(query)
else
  []
end
```

If you're working in a search form, you might want to create a
`search_form_changeset` and validate the search inputs before querying the
database.

### Using `CPF.Ecto.Changeset`

If you don't want use type in your CPF fields, you can use
`CPF.Ecto.Changeset.validate_cpf/2` function to validate a field in your
changeset. Example:

```elixir
CPF.Ecto.Changeset.validate_cpf(changeset, :cpf)
```

Using this approach you need manually cast the CPF field to store in uniform
way. As example, you can do that using the `Ecto.Change.prepare_changes/1`.
Example:

```elixir
changeset
|> CPF.Ecto.Changeset.validate_cpf(:cpf)
|> Ecto.Changeset.prepare_changes(fn changeset ->
  if input = Ecto.Changeset.get_change(changeset, :cpf) do
    string_cpf = input |> CPF.parse!() |> to_string()
    Ecto.Changeset.put_change(changeset, :cpf, string_cpf)
  else
    changeset
  end
end)
```

The `prepare_changes/2` will be called after the validation and before insert or
update in the repository. It gives the developer the opportunity to transform
the CPF value in any way they want.

## Why not other libraries?

This library runs 3 times faster and consume 3 times less memory and work with
primitive types, no extra struct is necessary.

## Docs

The docs can be found at [https://hexdocs.pm/cpf](https://hexdocs.pm/cpf).


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.2.0] - 2022-12-07
### Added
- Official support from Elixir 1.11 to Elixir 1.14. Thanks [@ayrat555](https://github.com/ayrat555)

### Removed

- Dropped support for Elixir versions below 1.11

## [1.1.0] - 2020-04-18

## Added

- Add official support to Elixir 1.10
- Add official support to Ecto 3.4 (cpf type in embed schemas)

## [1.0.0] - 2019-10-31

### Added
- Add `CPF.Ecto.Type` feature
- Add `CPF.Ecto.Changeset` feature

### Removed

- Removed support for wrong arguments in `CPF.valid?/1`
- Deleted `CPF.cast/1` and `CPF.cast!/1` functions
- Removed Elixir 1.5 support

## [0.8.1] - 2019-08-24
### Fix
- Mix tasks examples
- Visibility of the public mix tasks functions (they should be private)

## [0.8.0] - 2019-08-24
### Added
- Added `mix cpf.gen` and `mix cpf.check` tasks

## [0.7.1] - 2019-07-28
### Fix
- Fix documentation examples

## [0.7.0] - 2019-07-28
### Added
- `CPF.generate/0`

## [0.6.0] - 2019-07-07
### Added
- `CPF.flex/1`

## [0.5.1] - 2019-06-30

### Fixed
- Documentation examples

## [0.5.0] - 2019-06-30
### Added
- `to_string/1` and `CPF.to_integer/1`

### Deprecated
- Deprecated `CPF.cast/1` and `CPF.cast!/1` to favor `CPF.parse/1` and `CPF.parse!/1`

## [0.4.0] - 2019-06-16
### Added
- Add `CPF.cast/1` and `CPF.cast!/1` feature

### Deprecated
- Deprecated `CPF.valid?/1` with invalid types

## [0.3.0] - 2019-06-09
### Added
- Add `CPF.new/1` and `CPF.format/1` feature

### Deprecated
- Deprecated Elixir `1.5.0`

## [0.2.1] - 2019-04-29
### Fixed
- Make the CPF's Elixir version more forgiven

## [0.2.0] - 2019-04-21
### Added
- Add `CPF.valid?/1` for unformatted and formatted strings by [@lucaslvs](https://github.com/lucaslvs)

## [0.1.0] - 2019-04-14
### Added
- Add `CPF.valid?/1` for integers


lib/cpf/cpf.ex
defmodule CPF do
  @moduledoc """
  CPF module that provides functions to verify if a CPF is valid.
  """

  unless Version.match?(System.version(), ">= 1.7.0") do
    IO.warn("""
    You're running an old Elixir version. CPF will soon support Elixir versions
    above 1.7.0.
    """)
  end

  defstruct [:digits]

  defguardp is_pos_integer(number) when is_integer(number) and number >= 0

  @typep digit :: pos_integer()

  @typedoc """
  A text in `String.t()` or a positive integer
  """
  @type input :: String.t() | pos_integer

  @typedoc """
  The CPF type. It' composed of eleven digits(0-9]).
  """
  @opaque t :: %__MODULE__{
            digits: {digit, digit, digit, digit, digit, digit, digit, digit, digit, digit, digit}
          }

  @doc """
  Initializes a CPF.

  ## Examples

      iex> CPF.new(563_606_676_73)
      #CPF<563.606.676-73>

      iex> CPF.new("56360667673")
      #CPF<563.606.676-73>

  This function doesn't check if CPF numbers are valid, only use this function
  if the given `String.t` or the integer was validated before.
  """
  @spec new(input) :: t
  def new(valid_cpf) when is_pos_integer(valid_cpf) do
    %__MODULE__{
      digits: to_digits(valid_cpf)
    }
  end

  def new(valid_cpf) when is_binary(valid_cpf) do
    %__MODULE__{
      digits: valid_cpf |> String.to_integer() |> to_digits()
    }
  end

  @doc """
  Checks if the given argument is `CPF.t()` type.

  ## Examples

      iex> CPF.cpf?(563_606_676_73)
      false

      iex> "56360667673" |> CPF.new() |> CPF.cpf?()
      true
  """
  @spec cpf?(any) :: true | false
  def cpf?(%CPF{}), do: true
  def cpf?(_), do: false

  @doc """
  Returns `true` the given `cpf` is valid, otherwise `false`.

  ## Examples

      iex> CPF.valid?(563_606_676_73)
      true

      iex> CPF.valid?(563_606_676_72)
      false

      iex> CPF.valid?("563.606.676-73")
      true

      iex> CPF.valid?("563/60.6-676/73")
      false

      iex> CPF.valid?("563.606.676-72")
      false

      iex> CPF.valid?("56360667673")
      true

      iex> CPF.valid?("56360667672")
      false
  """
  @spec valid?(input :: input) :: boolean
  def valid?(input) when is_pos_integer(input) or is_binary(input) do
    case parse(input) do
      {:ok, _cpf} -> true
      {:error, _reason} -> false
    end
  end

  @doc """
  Returns a formatted string from a given  `cpf`.

  ## Examples

      iex> 563_606_676_73 |> CPF.new() |> CPF.format()
      "563.606.676-73"
  """
  @spec format(cpf :: t()) :: String.t()
  def format(%CPF{digits: digits}) do
    {dig_1, dig_2, dig_3, dig_4, dig_5, dig_6, dig_7, dig_8, dig_9, dig_10, dig_11} = digits

    to_string(dig_1) <>
      to_string(dig_2) <>
      to_string(dig_3) <>
      "." <>
      to_string(dig_4) <>
      to_string(dig_5) <>
      to_string(dig_6) <>
      "." <>
      to_string(dig_7) <>
      to_string(dig_8) <> to_string(dig_9) <> "-" <> to_string(dig_10) <> to_string(dig_11)
  end

  @doc """
  Builds a CPF struct by validating its digits and format. Returns an ok/error tuple for
  valid/invalid CPFs.

  ## Examples

      iex> {:ok, cpf} = CPF.parse(563_606_676_73)
      iex> cpf
      #CPF<563.606.676-73>

      iex> CPF.parse(563_606_676_72)
      {:error, %CPF.ParsingError{reason: :invalid_verifier}}
  """
  @spec parse(input) :: {:ok, t()} | {:error, CPF.ParsingError.t()}
  def parse(input) when is_pos_integer(input) do
    digits = Integer.digits(input)

    with {:ok, digits} <- add_padding(digits),
         {:ok, digits} <- skip_same_digits(digits),
         {:ok, digits} <- verify_digits(digits) do
      {:ok, %CPF{digits: digits}}
    end
  end

  def parse(
        <<left_digits::bytes-size(3)>> <>
          "." <>
          <<middle_digits::bytes-size(3)>> <>
          "." <>
          <<right_digits::bytes-size(3)>> <>
          "-" <>
          <<verifier_digits::bytes-size(2)>>
      ) do
    parse(left_digits <> middle_digits <> right_digits <> verifier_digits)
  end

  def parse(input) when is_binary(input) do
    case Integer.parse(input) do
      {int_input, ""} ->
        parse(int_input)

      _ ->
        {:error, %CPF.ParsingError{reason: :invalid_format}}
    end
  end

  @doc """
  Builds a CPF struct by validating its digits and format. Returns an CPF type
  or raises an `CPF.ParsingError` exception.

  ## Examples

      iex> CPF.parse!(563_606_676_73)
      #CPF<563.606.676-73>

      iex> CPF.parse!(563_606_676_72)
      ** (CPF.ParsingError) invalid_verifier
  """
  @spec parse!(input) :: t()
  def parse!(input) do
    case parse(input) do
      {:ok, cpf} -> cpf
      {:error, exception} -> raise exception
    end
  end

  @doc """
  Returns a tuple with the eleven digits of the given cpf.
  """
  @spec digits(t) :: tuple
  def digits(%CPF{digits: digits}), do: digits

  @doc """
  Returns a integer representation of the given cpf.

  ## Examples

      iex> 4_485_847_608 |> CPF.new() |> CPF.to_integer()
      4_485_847_608
  """
  @spec to_integer(t) :: pos_integer
  def to_integer(%CPF{digits: digits}) do
    0..10
    |> Enum.reduce(0, fn i, sum ->
      :math.pow(10, 10 - i) * elem(digits, i) + sum
    end)
    |> trunc()
  end

  @doc """
  Cleans up all characters of a given input except numbers. It can make CPF
  validation more flexbile.

  ## Examples

      iex> CPF.flex("  04.4 .8*58().476-08  ")
      "04485847608"

      iex> "  04.4 .8*58().476-08  " |> CPF.flex() |> CPF.valid?()
      true
  """
  @spec flex(String.t()) :: String.t()
  def flex(input) when is_binary(input), do: cleanup(input, "")

  defp cleanup("", cleaned), do: String.reverse(cleaned)

  defp cleanup(<<char::utf8, rest::binary>>, cleaned) when char in ?0..?9,
    do: cleanup(rest, <<char::utf8>> <> cleaned)

  defp cleanup(<<_char::utf8, rest::binary>>, cleaned), do: cleanup(rest, cleaned)

  @doc """
  Generates a random valid CPF.

  ## Examples

      iex> CPF.generate() |> to_string() |> CPF.valid?
      true
  """
  @spec generate :: t()
  def generate, do: gen()

  @doc """
    Generates a predictable random valid CPF wit the given `seed`.

  ## Examples

      iex> seed = {:exrop, [40_738_532_209_663_091 | 74_220_507_755_601_615]}
      iex> seed |> CPF.generate() |> CPF.format()
      "671.835.731-68"
  """
  @spec generate(seed :: :rand.builtin_alg() | :rand.state() | :rand.export_state()) :: t()
  def generate(seed) do
    :rand.seed(seed)
    gen()
  end

  defp gen do
    digits = {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0}

    digits =
      Enum.reduce(0..8, digits, fn index, digits ->
        put_elem(digits, index, Enum.random(0..9))
      end)

    {v1, v2} = calculate_verifier_digits(digits)

    digits =
      digits
      |> put_elem(9, v1)
      |> put_elem(10, v2)

    %CPF{digits: digits}
  end

  defp add_padding(digits) do
    padding = 11 - length(digits)

    if padding >= 0 do
      {:ok, add_padding(digits, padding)}
    else
      {:error, %CPF.ParsingError{reason: :too_long}}
    end
  end

  defp add_padding(digits, 0), do: digits

  defp add_padding(digits, padding) do
    add_padding([0 | digits], padding - 1)
  end

  defp skip_same_digits(digits) do
    if digits |> Enum.uniq() |> length() == 1 do
      {:error, %CPF.ParsingError{reason: :same_digits}}
    else
      {:ok, digits}
    end
  end

  defp verify_digits(digits) do
    cpf_digits = List.to_tuple(digits)
    {v1, v2} = calculate_verifier_digits(cpf_digits)
    {given_v1, given_v2} = extract_given_verifier_digits(cpf_digits)

    if v1 == given_v1 && v2 == given_v2 do
      {:ok, cpf_digits}
    else
      {:error, %CPF.ParsingError{reason: :invalid_verifier}}
    end
  end

  defp calculate_verifier_digits(digits) do
    {v1_sum, v2_sum} = sum_digits(digits)

    v1 = digit_verifier(v1_sum)
    v2_sum = v2_sum + 2 * v1
    v2 = digit_verifier(v2_sum)

    {v1, v2}
  end

  defp sum_digits(digits) do
    v1_weight = 10
    v2_weight = 11

    Enum.reduce(0..8, {0, 0}, fn i, {v1_sum, v2_sum} ->
      v1 = (v1_weight - i) * elem(digits, i)
      v2 = (v2_weight - i) * elem(digits, i)

      {v1_sum + v1, v2_sum + v2}
    end)
  end

  defp digit_verifier(sum) do
    rem = rem(sum, 11)
    if rem in [0, 1], do: 0, else: 11 - rem
  end

  defp extract_given_verifier_digits(digits) do
    {elem(digits, 9), elem(digits, 10)}
  end

  defp to_digits(integer) do
    digits = Integer.digits(integer)
    padding = 11 - length(digits)

    if padding >= 0 do
      digits
      |> add_padding(padding)
      |> List.to_tuple()
    else
      raise ArgumentError, "it has more than 11 digits"
    end
  end

  defimpl Inspect do
    import Inspect.Algebra

    def inspect(cpf, opts) do
      formatted_cpf = cpf |> CPF.format() |> color(:atom, opts)
      concat(["#CPF<", formatted_cpf, ">"])
    end
  end

  defimpl String.Chars do
    @doc """
    Returns a `String.t` representation of the given `cpf`.

    ## Examples

        iex> "04485847608" |> CPF.new() |> to_string()
        "04485847608"
    """
    @spec to_string(CPF.t()) :: String.t()
    def to_string(cpf) do
      digits = CPF.digits(cpf)
      for i <- 0..10, into: "", do: digits |> elem(i) |> Kernel.to_string()
    end
  end
end
